modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs

```diff
@@ -0,0 +1,388 @@
+                .Append("\", usage: (Godot.PropertyUsageFlags)")
+                .Append((int)propertyInfo.Usage)
+                .Append(", exported: ")
+                .Append(propertyInfo.Exported ? "true" : "false")
+                .Append(")");
+        }
+
+        private static MethodInfo DetermineMethodInfo(GodotMethodData method)
+        {
+            PropertyInfo returnVal;
+
+            if (method.RetType != null)
+            {
+                returnVal = DeterminePropertyInfo(method.RetType.Value, name: string.Empty);
+            }
+            else
+            {
+                returnVal = new PropertyInfo(VariantType.Nil, string.Empty, PropertyHint.None,
+                    hintString: null, PropertyUsageFlags.Default, exported: false);
+            }
+
+            int paramCount = method.ParamTypes.Length;
+
+            List<PropertyInfo>? arguments;
+
+            if (paramCount > 0)
+            {
+                arguments = new(capacity: paramCount);
+
+                for (int i = 0; i < paramCount; i++)
+                {
+                    arguments.Add(DeterminePropertyInfo(method.ParamTypes[i],
+                        name: method.Method.Parameters[i].Name));
+                }
+            }
+            else
+            {
+                arguments = null;
+            }
+
+            return new MethodInfo(method.Method.Name, returnVal, MethodFlags.Default, arguments,
+                defaultArguments: null);
+        }
+
+        private static PropertyInfo DeterminePropertyInfo(MarshalType marshalType, string name)
+        {
+            var memberVariantType = MarshalUtils.ConvertMarshalTypeToVariantType(marshalType)!.Value;
+
+            var propUsage = PropertyUsageFlags.Default;
+
+            if (memberVariantType == VariantType.Nil)
+                propUsage |= PropertyUsageFlags.NilIsVariant;
+
+            return new PropertyInfo(memberVariantType, name,
+                PropertyHint.None, string.Empty, propUsage, exported: false);
+        }
+
+        private static void GenerateHasMethodEntry(
+            string methodName,
+            StringBuilder source,
+            bool isFirstEntry
+        )
+        {
+            source.Append("        ");
+            if (!isFirstEntry)
+                source.Append("else ");
+            source.Append("if (method == GodotInternal.MethodName_");
+            source.Append(methodName);
+            source.Append(") {\n           return true;\n        }\n");
+        }
+
+        private static void GenerateMethodInvoker(
+            GodotMethodData method,
+            StringBuilder source
+        )
+        {
+            string methodName = method.Method.Name;
+
+            source.Append("        if (method == GodotInternal.MethodName_");
+            source.Append(methodName);
+            source.Append(" && argCount == ");
+            source.Append(method.ParamTypes.Length);
+            source.Append(") {\n");
```

Comment:
Maybe check the `arg.Type` here? Otherwise if you had an overload with the same amount of parameters but different types it will always go through the first one.

Reply:
We don't support that (neither now nor before this PR). It would be too slow to check this and error-prone (whatever Variant is supposed to pick). Godot's scripting API is just not made with this in mind.

Reply:
I thought since `godot_variant arg` has a `Variant.Type Type` property and `GodotMethodData method` has a `ImmutableArray<MarshalType> ParamTypes` property that maybe we could check, although we'd have to call `MarshalUtils.ConvertMarshalTypeToVariantType` on each param type (this call would be done by the generator, not in runtime).

My concern is if someone had something like this:

```csharp
void PickupItem(Item item); // Assuming Item is a Godot.Resource
void PickupItem(string itemId);
```

I know this is currently not supported but I was wondering if it could be in the future (not necessarily in this PR). Otherwise, wouldn't it be better to only add the invoker once? Because right now it is added for each overload even though only the first one will go through. This is what it currently generates for an overload:

```csharp
if (method == GodotInternal.MethodName_PickupItem && argCount == 1) {
    PickupItem(global::Godot.NativeInterop.VariantUtils.ConvertToStringObject(args[0]));
    ret = default;
    return true;
}
if (method == GodotInternal.MethodName_PickupItem && argCount == 1) {
    PickupItem((Item)global::Godot.NativeInterop.VariantUtils.ConvertToGodotObject(args[0]));
    ret = default;
    return true;
}
```

And we should consider adding an analyzer diagnostic if this is not supported.

Reply:
The problem is that a Variant is still convertible to a supported type even if the Variant type doesn't match. This means we would have to check for all possible conversions like Variant does, which is slower. There's also the issue about priority: should Variant.Type.Int choose the bool or the char overload? (or Variant.Array with its multiple supported types).

You're right we should not generate more than one overload with the same argument count, but I'm not sure about showing an error. It could be annoying if the user doesn't care about it (it would be different if method exposing was manual, with attributes).

Reply:
Is it possible to show an error only on usage instead of declaration? For example `node.Call("MyMethod")` when `MyMethod` could be ambiguous, or is that too costly?

Reply:
Yes, I think that would be the best option. But we need to add an attribute to help disambiguate. Because forcing the coder to refactor their method is not very nice. Something like an explicit `[Expose]` or `[GodotMethod]` (I'm not good at thinking names) to tell the source generator which one to pick.

Reply:
I'm thinking about this a bit more. Maybe we can attempt to resolve method overloads with the same parameter count by checking `Variant.Type` as you said, and error out if there's ambiguity (e.g.: `Node[]` and `Array` are both `Type.Array`, `int` and `byte` are both `Type.Int`). It would be strict, so `Type.Bool` cannot be assigned to `bool` and you can guess for others. Or maybe we can be a bit more permissive with some types.
The main problem is still the fact that showing errors can be annoying if the coder didn't mean to expose those methods to Godot.

Reply:
Of course, all this talk about method overloading is only relevant to scripts. In the GDExtension system we don't control this as it's just another ClassDB class. In fact, I think it's not possible at all to have overloads in a GDExtension class, as it's not possible to have them in an engine class. :eyes:

Reply:
I'm not sure forcing the user to add an attribute to their methods is better than forcing them to add a suffix, but if we use an attribute do you think it'd be a good idea to reuse the `[Export]` attribute? Probably not, and `[Expose]` seems too similar so I think it'd be confusing, maybe `[GodotMethod]` is the best option but I'm not good a naming either :laughing:.

Since GDScript doesn't support method overloading, I guess it makes sense that ClassDB doesn't have a way to bind them. I suppose GDExtension classes can have overloads as long as they are not bound in ClassDB or bound with different names.

Reply:
For GDExtension classes we will probably want method exposing to be opt-in with attributes.

Reply:
This will need more thought, so I'll be leaving it outside the scope of this PR. For now, I've made the generator only include the first overload with the same parameter count, so that it works like the old Mono implementation.